### ISSUE TITLE
False Positive with function call inside an array Index operation

### ISSUE THREAD

**Issue by HeroicKatora** [rust-lang/rust-clippy#7412]
Lint name: `clippy::unnecessary_operation`

I tried this code:

```rust
pub fn Int(x: usize) -> usize { x }

pub fn test(x: usize) {
    [0u8; 0][Int(x)];
}
```

I expected to see this happen: No warnings, or a warning about uncondition panic at most.

Instead, this happened:

```
Checking playground v0.0.1 (/playground)
warning: statement can be reduced
 --> src/lib.rs:4:5
  |
4 |     [0u8; 0][Int(x)];
  |     ^^^^^^^^^^^^^^^^^^^ help: replace it with: `[0u8; 0];Int(x);`
  |
  = note: `#[warn(clippy::unnecessary_operation)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_operation
```

The suggested transformation is wrong because the resulting code no longer panics for out-of-bounds indices.

### Meta

- `cargo clippy -V`: clippy 0.1.52 (9bc8c42 2021-05-09)
- `rustc -Vv`: 
  ```
  rustc 1.52.1 (9bc8c42bb 2021-05-09)
  binary: rustc
  commit-hash: 9bc8c42bb2f19e745a63f3445f1ac248fb015e53
  commit-date: 2021-05-09
  host: x86_64-unknown-linux-gnu
  release: 1.52.1
  LLVM version: 12.0.0
  ```
